pkg/prometheus/resource_selector.go

```diff
@@ -584,3 +587,68 @@ func validateProberURL(url string) error {
 	}
 	return nil
 }
+
+// SelectScrapeConfigs selects ScrapeConfigs based on the selectors in the Prometheus CR and filters them
+// returning only those with a valid configuration.
+func (rs *ResourceSelector) SelectScrapeConfigs(ctx context.Context) (map[string]*monitoringv1alpha1.ScrapeConfig, error) {
+	cpf := rs.p.GetCommonPrometheusFields()
+	objMeta := rs.p.GetObjectMeta()
+	namespaces := []string{}
+	// Selectors might overlap. Deduplicate them along the keyFunc.
+	scrapeConfigs := make(map[string]*monitoringv1alpha1.ScrapeConfig)
+
+	sConSelector, err := metav1.LabelSelectorAsSelector(cpf.ScrapeConfigSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	// If 'ScrapeConfigSelector' is nil only check own namespace.
+	if cpf.ScrapeConfigSelector == nil {
+		namespaces = append(namespaces, objMeta.GetNamespace())
+	} else {
+		sConNSSelector, err := metav1.LabelSelectorAsSelector(cpf.ScrapeConfigNamespaceSelector)
+		if err != nil {
+			return nil, err
+		}
+
+		namespaces, err = operator.ListMatchingNamespaces(sConNSSelector, rs.namespaceInformers)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	level.Debug(rs.l).Log("msg", "filtering namespaces to select ScrapeConfigs from", "namespaces", strings.Join(namespaces, ","), "namespace", objMeta.GetNamespace(), "prometheus", objMeta.GetName())
+
+	for _, ns := range namespaces {
+		err := rs.scrapeConfigLister.ListAllByNamespace(ns, sConSelector, func(obj interface{}) {
+			if k, ok := rs.accessor.MetaNamespaceKey(obj); ok {
+				scrapeConfig := obj.(*monitoringv1alpha1.ScrapeConfig).DeepCopy()
+				if err := k8sutil.AddTypeInformationToObject(scrapeConfig); err != nil {
+					level.Error(rs.l).Log("msg", "failed to set ScrapeConfig type information", "namespace", ns, "err", err)
+					return
+				}
+				scrapeConfigs[k] = scrapeConfig
+			}
+		})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list ScrapeConfigs in namespace %s", ns)
+		}
+	}
+
+	// TODO(xiu): add validation steps
+	var rejected int
+	res := make(map[string]*monitoringv1alpha1.ScrapeConfig, len(scrapeConfigs))
+
+	scrapeConfigKeys := make([]string, 0)
+	for k := range res {
+		scrapeConfigKeys = append(scrapeConfigKeys, k)
+	}
+	level.Debug(rs.l).Log("msg", "selected ScrapeConfigs", "scrapeConfig", strings.Join(scrapeConfigKeys, ","), "namespace", objMeta.GetNamespace(), "prometheus", objMeta.GetName())
+
+	if sKey, ok := rs.accessor.MetaNamespaceKey(rs.p); ok {
+		rs.metrics.SetSelectedResources(sKey, monitoringv1alpha1.ScrapeConfigsKind, len(res))
+		rs.metrics.SetRejectedResources(sKey, monitoringv1alpha1.ScrapeConfigsKind, rejected)
+	}
+
+	return res, nil
```

Comment:
I don't really get what you're doing here 😬 

You're creating an empty map `res`,  looping over this empty map and therefore not logging the selected keys nor returning anything 😅.

I'm not sure that the suggestion below works, but I guess it should be something like this?:
```suggestion
	// TODO(xiu): add validation steps
	var rejected int
	res := make(map[string]*monitoringv1alpha1.ScrapeConfig, len(scrapeConfigs))

	scrapeConfigKeys := make([]string, 0)
	for k := range scrapeConfigs {
		scrapeConfigKeys = append(scrapeConfigKeys, k)
	}
	level.Debug(rs.l).Log("msg", "selected ScrapeConfigs", "scrapeConfig", strings.Join(scrapeConfigKeys, ","), "namespace", objMeta.GetNamespace(), "prometheus", objMeta.GetName())

	if sKey, ok := rs.accessor.MetaNamespaceKey(rs.p); ok {
		rs.metrics.SetSelectedResources(sKey, monitoringv1alpha1.ScrapeConfigsKind, len(res))
		rs.metrics.SetRejectedResources(sKey, monitoringv1alpha1.ScrapeConfigsKind, rejected)
	}

	return scrapeConfigs, nil
```

Reply:
Haha, I'm not sure either 😅 This part was mostly an adaptation from Probe, will fix that, thanks!

Reply:
I actually cleaned up more than that, this was all needed with a validation logic but in our case we don't have one. See 7a9402af3